grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy

```diff
@@ -603,27 +605,28 @@ class ProductAvailabilityService {
         return quantityMap
     }
 
-    List getQuantityOnHandByBinLocation(Location location) {
-        def data = []
-
-        if (location) {
-            def results = ProductAvailability.executeQuery("""
-						select 
-						    pa.product, 
-						    pa.inventoryItem,
-						    pa.binLocation,
-						    sum(pa.quantityOnHand),
-						    sum(case when pa.quantityAvailableToPromise > 0 then pa.quantityAvailableToPromise else 0 end)
-						from ProductAvailability pa
-						left outer join pa.inventoryItem ii
-						left outer join pa.binLocation bl
-						where pa.location = :location
-						group by pa.product, pa.inventoryItem, pa.binLocation
-						""", [location: location])
+    List getQuantityOnHandByBinLocation(Location location, List<InventoryItem> inventoryItems = []) {
+        if (!location) {
+            throw new RuntimeException("Must specify location in order to get quantity on hand")
+        }
 
-            data = collectQuantityOnHandByBinLocation(results)
+        List results = ProductAvailability.createCriteria().list {
+            createAlias('inventoryItem', 'inventoryItem', JoinType.LEFT_OUTER_JOIN)
+            createAlias('binLocation', 'binLocation', JoinType.LEFT_OUTER_JOIN)
+            projections {
+                groupProperty("product", "product")
+                groupProperty("inventoryItem", "inventoryItem")
+                groupProperty("binLocation", "binLocation")
+                sum("quantityOnHand", "quantityOnHand")
+                sum("quantityAvailableToPromise", "quantityAvailableToPromise")
```

Comment:
have you made sure this would eventually return 0? I'm asking because of this:
```sql
sum(case when pa.quantityAvailableToPromise > 0 then pa.quantityAvailableToPromise else 0 end)
```

Reply:
a function that transforms this result `collectQuantityOnHandByBinLocation` handles the logic for setting a 0 for null values, but does not handle negative values...
Thanks for pointing this out, I think I am going to have to revert this to a simple SQL because it is quite hard to implement this logic with `CASE WHEN` in the criteria